## Exercise 1: Sensitivity Analyses

### Problem Statement:
What is the optimal ad placement policy that maximizes the click-through revenues while meeting the contractual obligations?

### Define decision variables:
Let $i={1,2,3,4}$ be in the index for the advertisers such that
* $i=1$ is GEICO
* $i=2$ is Delta
* $i=3$ is T-Mobile
* $i=4$ is Capital One

Let $j={1,2}$ be the index for the category, such as $j=1$ is the Sports categorty and $j=2$ is the National category.

So $x_{11}$ is the number of impressions for Geico in Sports and $x_{12}$ is the number of impressions for GEICO in the National category.

### Define objective function:
The objective of the platform, i.e., Washington Post, is to maximize advertising revenues. These revenues are driven by the cost per click charged to the advertisers, i.e., \$2.4 and the click-through rates.

Let $\kappa_{ij}$ be the CTR of advertiser $i$ in category $j$. For instance, the CTR of T-Mobile in the sports category is $\kappa_{31}=1.0\%$, whereas its CTR in the National category is $\kappa_{32}=3.0\%$.

The objective function is thus:
\begin{equation}
Objective=2.4\times \sum_{i=1}^4\sum_{j=1}^{2}  \kappa_{ij}\times x_{ij}
\end{equation}

### Define constraints:
The first table provides the following constraints

\begin{align}
x_{11}&\geq 2,000,000\\
x_{12}&\geq 1,000,000\\
x_{22}&\geq 1,000,000\\
x_{21}+x_{22}&\geq 2,000,000\\
x_{31}&\geq 1,000,000\\
x_{32}&\geq 1,000,000\\
x_{31}+x_{32}&\geq 3,000,000\\
x_{41}+x_{42}&\geq 2,000,000\\
\end{align}

"Assume that the Sports section gets six million views per day and the National section
get five million views per day"
This statement gives two constraints:
\begin{align}
\sum_{i=1}^{4}x_{i1}&\leq 6,000,000\\
\sum_{i=1}^{4}x_{i2}&\leq 5,000,000\\
\end{align}

Non-negativity constraints are
\begin{equation}
x_{ij}\geq 0
\end{equation}

In [0]:
import numpy 
import math
#!pip install cvxpy
import cvxpy as cvx
from numpy import *
from cvxpy import *
import pandas as pd

def solver(kappa):
    kappa1=kappa[0]
    kappa2=kappa[1]
    x1=cvx.Int(4,1)
    x2=cvx.Int(4,1)

    #Objective Function
    Z1=sum_entries(kappa1*x1)*2.4/100
    Z2=sum_entries(kappa2*x2)*2.4/100
    objective = cvx.Maximize((Z1+Z2))
  
    #Constraints
    c1=(sum_entries(x1))<=6*1000000# Capacity on Sports
    c2=(sum_entries(x2))<=5*1000000# Capacity on Sports
    c3=x1[0]>=2*1000000 # Geico must have at least 2m in sports
    c4=x2[0]>=1*1000000 # Geico must have at least 1m in sports
    c5=x2[1]>=1*1000000 # Delta must have at least 1m in sports
    c6=x1[1]+x2[1]==2*1000000 # Delta's total # of impressions must be equal to 2m total
    c7=x1[2]>=1*1000000# Tmobile must have at least 1m impression in sports
    c8=x2[2]>=1*1000000# Tmobile must have at least 1m impression in national
    c9=x1[2]+x2[2]==3*1000000 # Tmobile's total # of impressions must be equal 3m total
    c10=x1[3]+x2[3]==2*1000000# Capital Obe's impression equal 2m
    c=[c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,x1>=0,x2>=0]
  
    prob = cvx.Problem(objective, c)
    result = prob.solve()
    print('The optimal revenue is ')
    print(prob.value)
    print('The optimal allocation in the sports category is')
    print(x1.value)
    print('The optimal allocation in the national category is')
    print(x2.value)

In [34]:
# Find the optimal values of the decision variables and the objective function
solver(kappa=matrix([[2.5,2.0,1.0,1.5],[0.8,1.0,3.0,2.0]]))

The optimal revenue is 
523199.99999759113
The optimal allocation in the sports category is
[[2999999.99999905]
 [ 999999.99995182]
 [1000000.00001714]
 [1000000.0000317 ]]
The optimal allocation in the national category is
[[1000000.0000004 ]
 [1000000.00004818]
 [1999999.99998286]
 [ 999999.9999683 ]]


In [35]:
import numpy as np
# Sensitivity analyses of the optimal solution to different click-through rates
# Define company name, section name, and the original value of CTRs
company = ['Geico', 'Delta', 'T-Mobile', 'Capital One']
section = ['Sports', 'National']
CTR = [[2.5,2.0,1.0,1.5],[0.8,1.0,3.0,2.0]]

# Specify different CTRs as given in the HW
CTR_Sensitivity = [[np.linspace(2.0,3.0,11), np.linspace(1.5,2.5,11), np.linspace(0.5,1.5,11), np.linspace(1.5,2.0,6)],
      [np.linspace(0.1,1.1,11), np.linspace(0.5,1.5,11), np.linspace(2.5,3.5,11), np.linspace(1.5,2.5,11)]]

# Run a for loop to implement sensitivity analyses one cell at a time
for i_section in range(len(section)):
    for i_company in range(len(company)):
        CTR_matrix = CTR
        for n in CTR_Sensitivity[i_section][i_company]:
            print(f'Sensitivity analyses with CTR of {company[i_company]}+{section[i_section]} = {n}%:')
            CTR_matrix[i_section][i_company] = n
            solver(kappa=matrix(CTR_matrix))
            print('\n')

Sensitivity analyses with CTR of Geico+Sports = 2.0%:
The optimal revenue is 
487199.9999981021
The optimal allocation in the sports category is
[[2999999.99999848]
 [ 999999.99998563]
 [1000000.00003637]
 [ 999999.99997915]]
The optimal allocation in the national category is
[[1000000.00000082]
 [1000000.00001437]
 [1999999.99996362]
 [1000000.00002085]]


Sensitivity analyses with CTR of Geico+Sports = 2.1%:
The optimal revenue is 
494399.99999717786
The optimal allocation in the sports category is
[[2999999.99999857]
 [ 999999.99998237]
 [1000000.00005878]
 [ 999999.9999599 ]]
The optimal allocation in the national category is
[[1000000.0000007 ]
 [1000000.00001763]
 [1999999.99994122]
 [1000000.0000401 ]]


Sensitivity analyses with CTR of Geico+Sports = 2.2%:
The optimal revenue is 
501599.9999970976
The optimal allocation in the sports category is
[[2999999.9999989 ]
 [ 999999.99998037]
 [1000000.00005939]
 [ 999999.99996102]]
The optimal allocation in the national category is
[[

Sensitivity analyses with CTR of Capital One+National = 2.1%:
The optimal revenue is 
640799.9999997189
The optimal allocation in the sports category is
[[2999999.99999873]
 [ 999999.99999664]
 [1000000.0000025 ]
 [1000000.00000181]]
The optimal allocation in the national category is
[[1000000.0000006 ]
 [1000000.00000336]
 [1999999.9999975 ]
 [ 999999.99999819]]


Sensitivity analyses with CTR of Capital One+National = 2.2%:
The optimal revenue is 
643199.999998521
The optimal allocation in the sports category is
[[2999999.99998905]
 [ 999999.99998705]
 [1000000.00000934]
 [1000000.00001132]]
The optimal allocation in the national category is
[[1000000.00000467]
 [1000000.00001295]
 [1999999.99999066]
 [ 999999.99998868]]


Sensitivity analyses with CTR of Capital One+National = 2.3%:
The optimal revenue is 
645599.9999990595
The optimal allocation in the sports category is
[[2999999.99998858]
 [ 999999.99999694]
 [ 999999.99999586]
 [1000000.00000264]]
The optimal allocation in the n

### Write no more than one paragraph (no more than 300 words) to describe the results of your analyses. Tables or graphs can be provided as supplemental material.<br>
- The optimal revenue increases as the click through rate increases. This intuitively make sense, cause the higer CTR will bring more views, and more views bring higher revenue.
- The optimal AD allocation in each category does not change with the change of click through rate. After all, the CTR only change in a small range, and this range is not large enough to make change on allocation plan.

## Exercise 2: Operations Excellence

## Plant A

### Problem Statement:
What is the optimal production plan for the two products, “standard” and “deluxe” that maximizes the company's profit.

### Define decision variables:
Let $i={a,b}$ be in the index for the plant A and plant B, respectively.
* $i=a$ is plant A
* $i=b$ is plant B

Let $j={s,d}$ be in the index for the standard and deluxe, respectively.
* $j=s$ is standard
* $j=d$ is deluxe

So $x_{as}$ is the number of standard products for plant A and $x_{ad}$ is the number of deluxe products for plant A.

### Define objective function:

\begin{equation}
Objective=10\times x_{as} + 15\times x_{ad}
\end{equation}

### Define constraints:
\begin{align}
4x_{as}+2x_{ad}&\leq 80\\
2x_{as}+5x_{ad}&\leq 60\\
4x_{as}+4x_{ad}&\leq 75\\
\end{align}

Non-negativity constraints are
\begin{align}
x_{as}&\geq 0\\
x_{ad}&\geq 0\\
\end{align}

In [36]:
import numpy 
import math
#!pip install cvxpy
import cvxpy as cvx
from numpy import *
from cvxpy import *
import pandas as pd

kappa=matrix([10,15])
x=cvx.Int(2,1)

#Objective Function
Z1=sum_entries(kappa*x)
objective = cvx.Maximize(Z1)
  
#Constraints
c1=4*x[0]+2*x[1]<=80 # Plant A has a grinding capacity of 80 hours per week
c2=2*x[0]+5*x[1]<=60 # Plant A has a polishing capacity of 60 hours per week
c3=4*x[0]+4*x[1]<=75 # Plant A is allocated 75 kg of raw per week
c=[c1,c2,c3,x>=0]
  
prob = cvx.Problem(objective, c)
result = prob.solve()
print('The optimal profit is')
print(prob.value)
print('The optimal allocation in the “standard” product is')
print(x[0].value)
print('The optimal allocation in the “deluxe” product is')
print(x[1].value)

The optimal profit is
220.00000005215267
The optimal allocation in the “standard” product is
10.000000000492811
The optimal allocation in the “deluxe” product is
8.000000003148305


## Plant B

### Problem Statement:
What is the optimal production plan for the two products, “standard” and “deluxe” that maximizes the company's profit.

### Define decision variables:
Let $i={a,b}$ be in the index for the plant A and plant B, respectively.
* $i=a$ is plant A
* $i=b$ is plant B

Let $j={s,d}$ be in the index for the standard and deluxe, respectively.
* $j=s$ is standard
* $j=d$ is deluxe

So $x_{bs}$ is the number of standard products for plant B and $x_{bd}$ is the number of deluxe products for plant B.

### Define objective function:

\begin{equation}
Objective=10\times x_{bs} + 15\times x_{bd}
\end{equation}

### Define constraints:
\begin{align}
5x_{bs}+3x_{bd}&\leq 60\\
5x_{bs}+6x_{bd}&\leq 75\\
4x_{bs}+4x_{bd}&\leq 45\\
\end{align}

Non-negativity constraints are
\begin{align}
x_{bs}&\geq 0\\
x_{bd}&\geq 0\\
\end{align}

In [37]:
import numpy 
import math
#!pip install cvxpy
import cvxpy as cvx
from numpy import *
from cvxpy import *
import pandas as pd

kappa=matrix([10,15])
x=cvx.Int(2,1)

#Objective Function
Z1=sum_entries(kappa*x)
objective = cvx.Maximize(Z1)
  
#Constraints
c1=5*x[0]+3*x[1]<=60 # Plant B has a grinding capacity of 60 hours per week
c2=5*x[0]+6*x[1]<=75 # Plant B has a polishing capacity of 75 hours per week
c3=4*x[0]+4*x[1]<=45 # Plant B is allocated 45 kg of raw per week
c=[c1,c2,c3,x>=0]

prob = cvx.Problem(objective, c)
result = prob.solve()
print('The optimal profit is')
print(prob.value)
print('The optimal allocation in the “standard” product is')
print(x[0].value)
print('The optimal allocation in the “deluxe” product is')
print(x[1].value)

The optimal profit is
165.00000006681978
The optimal allocation in the “standard” product is
2.912859137415921e-09
The optimal allocation in the “deluxe” product is
11.000000002512746


## Joint optimization of Plant A and B

### Problem Statement:
What is the optimal production plan for the two products, “standard” and “deluxe” that maximizes the company's profit.

### Define decision variables:
Let $i={a,b}$ be in the index for the plant A and plant B, respectively.
* $i=a$ is plant A
* $i=b$ is plant B

Let $j={s,d}$ be in the index for the standard and deluxe, respectively.
* $j=s$ is standard
* $j=d$ is deluxe

So $x_{as}$ is the number of standard products for plant A and $x_{bd}$ is the number of deluxe products for plant B.

### Define objective function:

\begin{equation}
Objective=10\times x_{as} + 15\times x_{ad} + 10\times x_{bs} + 15\times x_{bd}
\end{equation}

### Define constraints:
\begin{align}
4x_{as}+2x_{ad}&\leq 80\\
2x_{as}+5x_{ad}&\leq 60\\
5x_{bs}+3x_{bd}&\leq 60\\
5x_{bs}+6x_{bd}&\leq 75\\
4x_{as}+4x_{ad}+4x_{bs}+4x_{bd}&\leq 120\\
\end{align}

Non-negativity constraints are
\begin{equation}
x_{ij}\geq 0
\end{equation}

In [38]:
import numpy 
import math
#!pip install cvxpy
import cvxpy as cvx
from numpy import *
from cvxpy import *
import pandas as pd

kappa=matrix([[10,15],[10,15]])
kappa1=kappa[0]
kappa2=kappa[1]
x1=cvx.Int(2,1)
x2=cvx.Int(2,1)

#Objective Function
Z1=sum_entries(kappa1*x1)
Z2=sum_entries(kappa2*x2)
objective = cvx.Maximize(Z1+Z2)
  
#Constraints
c1=4*x1[0]+2*x1[1]<=80 # Plant A has a grinding capacity of 80 hours per week
c2=2*x1[0]+5*x1[1]<=60 # Plant A has a polishing capacity of 60 hours per week
c3=5*x2[0]+3*x2[1]<=60 # Plant B has a grinding capacity of 60 hours per week
c4=5*x2[0]+6*x2[1]<=75 # Plant B has a polishing capacity of 75 hours per week
c5=4*x1[0]+4*x1[1]+4*x2[0]+4*x2[1]<=120 # Plant B and Plant B are allocated 120 kg of raw per week in total
c=[c1,c2,c3,c4,c5,x1>=0,x2>=0]

prob = cvx.Problem(objective, c)
result = prob.solve()
print('The optimal profit is')
print(prob.value)
print('Plant A - The optimal allocation in the “standard” product is')
print(x1[0].value)
print('Plant A - The optimal allocation in the “deluxe” product is')
print(x1[1].value)
print('Plant B - The optimal allocation in the “standard” product is')
print(x2[0].value)
print('Plant B - The optimal allocation in the “deluxe” product is')
print(x2[1].value)

The optimal profit is
399.99999992548265
Plant A - The optimal allocation in the “standard” product is
9.999999999202224
Plant A - The optimal allocation in the “deluxe” product is
7.9999999965348945
Plant B - The optimal allocation in the “standard” product is
-1.4098915363594164e-09
Plant B - The optimal allocation in the “deluxe” product is
11.999999999969063


### Please comment on the results? How do they compare to the results obtained in Question 2.<br>
The profit summation of plant A and plant B (220+165=385) is less than the profit of joint optimization of plant A and B(400). This happens because we set a limitation of row material on each plant. When we remove this limitation, we can now better take use of all(and allocate) row material. In other word, the optimal profit will increase with the decrease in limitation. In this case, we use the "wasted" row material from plant A, to add to the needed row material that producing additional one deluxe product in plant B, to reach higher joint profit. Hence, the additional 15 profit(400-385=15) is from one additional deluxe product producing in plant B.

### In one paragraph (no more than 300 words), detail the recommendations you would provide to the CEO of the company to improve the firm’s operations?
- Use joint production, do not set limitation of row material on each plant.
- Plant A - The optimal allocation in the “standard” product is 10. Plant A - The optimal allocation in the “deluxe” product is 8. Plant B - The optimal allocation in the “standard” product is 0 and Plant B - The optimal allocation in the “deluxe” product is 12.

## Exercise 3: Hotel La Quinta Motor Inns (LQM)

## Q1
### According to the regression equation given above, which variable positively affect Profitability? <br> Which variable negatively affect Profitability? Does this intuitively make sense? Why?<br>
Variables that are positively affect Profitability:
- Price of the Inn
- College Students in Area<br>

The number of students in area is positively related to the Profitability of Inns. This is intuitively make sense.<br>
But the price of the Inn is not intuitively necessarily positively related to the Profitability of Inns.<br><br>
Variables that are negatively affect Profitability:
- State Population Per Inn
- Square Root of Median Income<br>

State population per Inn is not intuitively necessairly negatively related to the Profitability of Inns.<br> 
The square root of median income is intuitively negatively related to the Profitability of Inns.

## Q2
### Using this regression equation, LQM created a spreadsheet model to predict profitability. LQM collected data for several locations in California, which is provided in the excel spreadsheet on Canvas “LQM”. Using this spreadsheet, compute the profitability for each hotel. Which one has the highest profitability? Which one has the lowest profitability?

In [51]:
from google.colab import files
files.upload()

Saving HW1_LQM.xlsx to HW1_LQM (1).xlsx


{'HW1_LQM.xlsx': b'PK\x03\x04\x14\x00\x06\x00\x08\x00\x00\x00!\x00;H\x8e@l\x01\x00\x00\xc4\x04\x00\x00\x13\x00\x08\x02[Content_Types].xml \xa2\x04\x02(\xa0\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\

In [40]:
#!pip install xlrd
# display the excel file
import pandas as pd
df = pd.read_excel('HW1_LQM.xlsx', header=1)
df['Profitability'] = 39.05-(5.41*df['State Population Per Inn (normalized)'])+(5.86*df['Price (normalized)'])-(3.09*df['Square Root of Median Income (normalized)'])+(1.75*df['College Students in Area (normalized)'])
df = df.sort_values(by=['Profitability'], ascending=False)
print('Hotels that have the highest profitability and the lowest profitability: \n')
df.iloc[[0,-1],[0,1,-1]]

Hotels that have the highest profitability and the lowest profitability: 



Hotel                 Location  Profitability
1      2       Fresno, California      53.379192
7      8  Los Angeles, California      23.445409

## Q3
### a. LQM has a budget of $10 million to spend on hotels. Suppose we used a “greedy” approach where we selected the most profitable hotels until we ran out of budget. So we would start by the most profitable, and then if we had enough budget left, we would buy the hotel we predict to be the second most profitable, and so on.

In [41]:
budget = 0
count = 0
for i in range(len(df)):
    if budget >= 10000000:
    break
    
    budget += df.iloc[i,2]
    print(f'Purchase hotel {df.iloc[i,0]}, {df.iloc[i,1]}')

Purchase hotel 2, Fresno, California


### b. What would our total predicted profitability be? (This is the sum of the predicted profitability of all hotels we purchase.)

In [42]:
print(f'The total predicted profitability = {df.Profitability[df.Hotel == 2].values[0]}')

The total predicted profitability = 53.379192308346


### c. If we are trying to maximize our total predicted profitability, is this a good approach? How about if we were trying to maximize the average predicted profitability of the hotels we select? How about if we had a budget of 20 million instead of 10 million?

Answer: Above is not a good approach to maximize the total predicted profitability. The hotel that has the highest profitability may cost too much. A better investment can be reached by using linear programming to find the allocation plan with maximum total profitability. From the profit point of view, maximizing the total profitability make more sense than maximizing the average predicted profitability. If the budget increases to 20 million, more hotels can be purchased.

## Q4
### Now, build an optimization model to select hotels given the $10 million budget.

### Problem Statement:
What is the total predicted profitability be? (the sum of the predicted profitability of all hotels we purchase.)

### Define decision variables:
Let $i={1,2,3,...,16}$ be in the index for the 16 hotels.
* $x1$ stands for hotel index 1
* $x2$ stands for hotel index 2
* The value of $xi$ are belong to 0(not purchase) or 1(purchase)

### Define objective function:
Let $\kappa_{i}$ be the profitability of hotels.<br>
The total number of hotels is represented as n, n=16

\begin{equation}
Objective=\sum_{i=1}^n \kappa_{i}\times x_{i}
\end{equation}<br>
\begin{align}
x_{i}\in&\{0,1\}&\\
\end{align}

### Define constraints:
Let $j_{i}$ be the price of hotels.<br>
\begin{align}
\sum_{i=1}^n j_{i}\times x_{i}\leq 10000000\\
\end{align}<br>
\begin{align}
x_{i}\in&\{0,1\}&\\
\end{align}

In [43]:
import numpy as np
import math
from numpy import matrix 
#!pip install cvxpy
from cvxpy import *
import cvxpy as cvx
import pandas as pd

N=len(df) # Number of possible hotels
d=matrix([list(df.Profitability),list(df.Price)]) # Profitability and price matrix

# Define variables
x = cvx.Variable(N) # Whether to purchase the hotels.
y = cvx.Bool(N) # Define whether the 16 hotels are going to be purchased or not.

Z0 = sum_entries(d[0]*x)
Z1 = sum_entries(d[1]*y)

# Construction of the Objective
objective = cvx.Maximize(Z0)

# Construction of the Constraints
## The total budget is limited to 10 million dollars
c0 = Z1 <= 10000000

## Selected hotels can deliver only if these hotels are selected to be purchased
c1 = x[0] <= y[0]
c2 = x[1] <= y[1]
c3 = x[2] <= y[2]
c4 = x[3] <= y[3]
c5 = x[4] <= y[4]
c6 = x[5] <= y[5]
c7 = x[6] <= y[6]
c8 = x[7] <= y[7]
c9 = x[8] <= y[8]
c10 = x[9] <= y[9]
c11 = x[10] <= y[10]
c12 = x[11] <= y[11]
c13 = x[12] <= y[12]
c14 = x[13] <= y[13]
c15 = x[14] <= y[14]
c16 = x[15] <= y[15]

## xs must be greater than or equal to zero
c17 = x >= 0

## set of constraints
con=[c0,c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13,c14,c15,c16,c17]

prob = cvx.Problem(objective, con)
result = prob.solve()

print('The maximum profitability is:')
print(prob.value)
print('Hotels to purchase are:')
print(x.value)

The maximum profitability is:
269.92468137325164
Hotels to purchase are:
[[1.30654410e-11]
 [1.25423038e-11]
 [1.18628957e-11]
 [1.16838358e-11]
 [1.16207611e-11]
 [1.15836760e-11]
 [1.00000000e+00]
 [1.00000000e+00]
 [1.00000000e+00]
 [1.00000000e+00]
 [1.00000000e+00]
 [1.00000000e+00]
 [1.00000000e+00]
 [9.05035346e-12]
 [7.87793912e-12]
 [7.79216163e-12]]


In [44]:
# Hotels to purchase are:
hotel_purchase = df.loc[pd.DataFrame(np.round(x.value)).iloc[:,0].values * df['Hotel'] > 0,['Hotel','Location','Price','Profitability']]
hotel_purchase

Hotel                      Location    Price  Profitability
11     12  South Lake Tahoe, California  2500000      40.289363
12     13  South Lake Tahoe, California  1975000      39.419290
9      10  South Lake Tahoe, California  1650000      38.880673
14     15  South Lake Tahoe, California  1475000      38.590649
10     11  South Lake Tahoe, California  1125000      38.010600
15     16  South Lake Tahoe, California   750000      37.389119
4       5            Fresno, California   325000      37.344988

### Does the optimal solution make sense intuitively? How does it compared to the greedy solution?
This optimal solution does make sense intuitively. Comparing to the greedy solution, this model better find out the total profitability given considering the price(cost) for each hotel. In the greedy solution, that model does not consider if the hotel is actually affordable. A hotel may have high profitability but it may not be the best buy. In the optimization model, we select the best hotel combination (maximize total profitability) that we can afford in the 10 million budget.

## Q5
### LQM thinks that buying too many hotels in one city is probably not a good idea and would prefer to diversify across as many cities as possible. Add constraint(s) to your model to limit the number of hotels purchased in any city to at most 2.

### a. What are the constraints that you need to add to the model? Intuitively, do you expect the new optimal objective function value to be larger, smaller or the same as before?
- Add constraint - limit the number of hotel to purchse in each city up to 2(see below LP definition for details).
- Intuitively, the new optimal objective function value will be smaller than before, because the new optimization problem has more limitation.

In [45]:
# sort data set by Location
df = df.sort_values(by=['Location'])
df.head()

Hotel            Location     Price  Price (normalized)  \
0      1  Eureka, California   2925000           -0.301823   
1      2  Fresno, California  10000000            1.699076   
2      3  Fresno, California   3750000           -0.068503   
3      4  Fresno, California   3500000           -0.139206   
4      5  Fresno, California    325000           -1.037136   

   Square Root of Median Income (normalized)  \
0                                  -0.812780   
1                                  -0.408199   
2                                  -0.408199   
3                                  -0.408199   
4                                  -0.408199   

   College Students in Area (normalized)  \
0                              -0.536413   
1                               0.311669   
2                               0.311669   
3                               0.311669   
4                               0.311669   

   State Population Per Inn (normalized)  Profitability  
0                              -0.995987      44.242369  
1                              -0.474279      53.379192  
2                              -0.474279      43.021179  
3                              -0.474279      42.606858  
4                              -0.474279      37.344988

In [46]:
# count number of hotels in each city
df.groupby('Location')['Location'].count()

Location
Eureka, California              1
Fresno, California              4
Long Beach, California          1
Los Angeles, California         3
South Lake Tahoe, California    7
Name: Location, dtype: int64

### Problem Statement:
What is the total predicted profitability be? (the sum of the predicted profitability of all hotels we purchase.)

### Define decision variables:
Let $i={1,2,3,...}$ be the index for different cities.<br>
Let $z={1,2,3,...}$ be the index for different hotels.<br>
* $xiz$ stands for the hotel with index z in city i
* $x21$ the first(index 1) hotel in city 2
* The value of $xiz$ are belong to 0(not purchase) or 1(purchase)

### Define objective function:
Let $\kappa_{iz}$ be the profitability of hotels.<br>
The total number of hotels in each city is represented as n.

\begin{equation}
Objective=\sum_{i=1}^5\sum_{z=1}^{n}  \kappa_{iz}\times x_{iz}
\end{equation}<br>
\begin{align}
x_{iz}\in&\{0,1\}&\\
\end{align}

### Define constraints:
Let $j_{iz}$ be the price of hotels.<br>
\begin{align}
\sum_{i=1}^5\sum_{z=1}^{n} j_{iz}\times x_{iz}\leq 10000000\\
\end{align}<br>

\begin{align}
\sum_{z=1}^{1} x_{1z}\leq 2\\
\end{align}<br>
\begin{align}
\sum_{z=1}^{4} x_{2z}\leq 2\\
\end{align}<br>
\begin{align}
\sum_{z=1}^{1} x_{3z}\leq 2\\
\end{align}<br>
\begin{align}
\sum_{z=1}^{3} x_{4z}\leq 2\\
\end{align}<br>
\begin{align}
\sum_{z=1}^{7} x_{5z}\leq 2\\
\end{align}<br>

\begin{align}
x_{iz}\in&\{0,1\}&\\
\end{align}

In [47]:
import numpy as np
import math
from numpy import matrix 
#!pip install cvxpy
from cvxpy import *
import cvxpy as cvx
import pandas as pd

N=len(df) # Number of possible hotels
d=matrix([list(df.Profitability),list(df.Price)]) # Profitability and price matrix

# Define variables
x = cvx.Variable(N) # Whether to purchase the hotels.
y = cvx.Bool(N) # Define whether the 16 hotels are going to be purchased or not.

Z0 = sum_entries(d[0]*x)
Z1 = sum_entries(d[1]*y)

# Construction of the Objective
objective = cvx.Maximize(Z0)

# Construction of the Constraints
## The total budget is limited to 10 million dollars
c0 = Z1 <= 10000000

## Selected hotels can deliver only if these hotels are selected to be purchased
c1 = x[0] <= y[0]
c2 = x[1] <= y[1]
c3 = x[2] <= y[2]
c4 = x[3] <= y[3]
c5 = x[4] <= y[4]
c6 = x[5] <= y[5]
c7 = x[6] <= y[6]
c8 = x[7] <= y[7]
c9 = x[8] <= y[8]
c10 = x[9] <= y[9]
c11 = x[10] <= y[10]
c12 = x[11] <= y[11]
c13 = x[12] <= y[12]
c14 = x[13] <= y[13]
c15 = x[14] <= y[14]
c16 = x[15] <= y[15]
c17 = sum_entries(x[0]) <= 2
c18 = sum_entries(x[1:5]) <= 2
c19 = sum_entries(x[5]) <= 2
c20 = sum_entries(x[6:9]) <= 2
c21 = sum_entries(x[9:]) <= 2

## xs must be greater than or equal to zero
c22 = x >= 0

## set of constraints
con=[c0,c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13,c14,c15,c16,c17,c18,c19,c20,c21,c22]

prob = cvx.Problem(objective, con)
result = prob.solve()

print('The maximum profitability is:')
print(prob.value)
print('Hotels to purchase are:')
print(x.value)

The maximum profitability is:
205.70090438609586
Hotels to purchase are:
[[1.00000000e+00]
 [1.84129037e-10]
 [1.62679871e-10]
 [1.61723176e-10]
 [1.00000000e+00]
 [1.46817624e-10]
 [1.00870831e-10]
 [9.99999999e-01]
 [9.99999999e-01]
 [9.17563796e-11]
 [8.84809353e-11]
 [8.72542768e-11]
 [1.00000000e+00]
 [8.61575768e-11]
 [1.00000000e+00]
 [8.46775618e-11]]


In [48]:
# Hotels to purchase are:
hotel_purchase2 = df.loc[pd.DataFrame(np.round(x.value)).iloc[:,0].values * df['Hotel'] > 0,['Hotel','Location','Price','Profitability']]
hotel_purchase2

Hotel                      Location    Price  Profitability
0       1            Eureka, California  2925000      44.242369
4       5            Fresno, California   325000      37.344988
6       7       Los Angeles, California  1950000      23.776866
7       8       Los Angeles, California  1750000      23.445409
9      10  South Lake Tahoe, California  1650000      38.880673
10     11  South Lake Tahoe, California  1125000      38.010600

### How does above compare to the previous solution?
- The new optimal objective function value is smaller than the previous solution(206 < 270), because the new optimization problem has more limitation. The limitation on number of hotels to purchase in each city make the model to find optimal solution in a limited range only.

### In one paragraph (no more than 300 words), describe how you would present your results to LQM. Do you have any recommendations for them to improve the regression model? How about to improve the optimization model?
- Recommendation on regression model: the relationship between profitability and variables may not be linear, a more complex model(such as nonlinear) could be tried in the future.
- Do not limit the number of hotels to purchase in each city if we consider the total profitability only. More variables could be added to model such as defining a variable to consider the influence on the number of hotels to purchase in a city.
- If we DO NOT limit the number of hotel to purchase in each city, the optimal purchase plan is:

In [49]:
hotel_purchase

Hotel                      Location    Price  Profitability
11     12  South Lake Tahoe, California  2500000      40.289363
12     13  South Lake Tahoe, California  1975000      39.419290
9      10  South Lake Tahoe, California  1650000      38.880673
14     15  South Lake Tahoe, California  1475000      38.590649
10     11  South Lake Tahoe, California  1125000      38.010600
15     16  South Lake Tahoe, California   750000      37.389119
4       5            Fresno, California   325000      37.344988

- If we DO limit the number of hotel to purchase in each city, the optimal purchase plan is:

In [50]:
hotel_purchase2

Hotel                      Location    Price  Profitability
0       1            Eureka, California  2925000      44.242369
4       5            Fresno, California   325000      37.344988
6       7       Los Angeles, California  1950000      23.776866
7       8       Los Angeles, California  1750000      23.445409
9      10  South Lake Tahoe, California  1650000      38.880673
10     11  South Lake Tahoe, California  1125000      38.010600